Lib/fontTools/otlLib/optimize.py

```diff
@@ -0,0 +1,451 @@
+            all_pairs, int(mode)
+        )
+        for pairs in grouped_pairs:
+            subtables.append(
+                buildPairPosClassesSubtable(pairs, font.getReverseGlyphMap())
+            )
+    else:
+        raise ValueError(f"Bad {GPOS_COMPACT_MODE_ENV_KEY}={mode}")
+    return subtables
+
+
+def is_really_zero(class2: otTables.Class2Record) -> bool:
+    v1 = getattr(class2, "Value1", None)
+    v2 = getattr(class2, "Value2", None)
+    return (v1 is None or v1.getEffectiveFormat() == 0) and (
+        v2 is None or v2.getEffectiveFormat() == 0
+    )
+
+
+Pairs = Dict[
+    Tuple[Tuple[str, ...], Tuple[str, ...]],
+    Tuple[otBase.ValueRecord, otBase.ValueRecord],
+]
+
+
+def cluster_pairs_by_class2_coverage_agglomerative(
+    pairs: Pairs, clustering_kwargs: Dict = {}
+) -> List[Pairs]:
+    from sklearn import cluster
+
+    # Idea:
+    #   1. group pairs by Class1 = get lines from the matrix
+    #   2. cluster together lines that have non-zero values in the same Class2
+    #   3. make 1 subtable per cluster
+    # That way, we hope to have fewer subtables than in the "max" strategy,
+    # but that each subtable will have high occupancy/low sparsity thanks to
+    # the clustering.
+    all_class1 = list(set(pair[0] for pair in pairs))
+    all_class2 = list(set(pair[1] for pair in pairs))
+    vectors = [
+        [1 if (class1, class2) in pairs else 0 for class2 in all_class2]
+        for class1 in all_class1
+    ]
+
+    if len(all_class1) <= 1:
+        # Can't cluster just 1 line
+        return [pairs]
+
+    # Make matrix lines
+    lines: Dict[Sequence[str], Pairs] = defaultdict(dict)
+    for pair, values in pairs.items():
+        lines[pair[0]][pair] = values
+    # https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html
+    # Default values determined by running:
+    # python -m scripts.measure_kerning_optimizations.find_best_clustering
+    kwargs = {**clustering_kwargs}  # Shallow copy to edit defaults
+    if "lines_per_cluster" in kwargs:
+        lines_per_cluster = kwargs.pop("lines_per_cluster")
+        kwargs["n_clusters"] = min(
+            max(int(len(lines) / lines_per_cluster), 1), len(lines)
+        )
+        # print(len(lines), kwargs["n_clusters"])
+        kwargs["distance_threshold"] = None
+    else:
+        kwargs.setdefault("n_clusters", None)
+        kwargs.setdefault("linkage", "ward")
+        kwargs.setdefault("distance_threshold", 5.5)
+    labels = cluster.AgglomerativeClustering(**kwargs).fit_predict(vectors)
+    # Group matrix lines according to clustering
+    grouped_lines: Dict[int, Pairs] = defaultdict(dict)
+    ungrouped_lines = []
+    for class1, label in zip(all_class1, labels):
+        if label == -1:
+            ungrouped_lines.append(lines[class1])
+        else:
+            grouped_lines[label].update(lines[class1])
+    return [*grouped_lines.values(), *ungrouped_lines]
+
+
+def _bit_count(n):
```

Comment:
Let's move this (and other useful facilities) into `fontTools.misc.bitTools`.

Reply:
Of.. `fontTools.misc.intTools`... which... already has a `popCount` which does the same. We should measure them maybe and pick one.

Reply:
Filed https://github.com/fonttools/fonttools/issues/2331